docs/customization.rst

```diff
@@ -325,40 +282,69 @@ Configuration Options
 
             To use light and dark modes, this theme supports 2 schemes which are specified by a ``scheme`` field.
 
-            1. The ``default`` scheme for light mode
+            .. code-block:: python
+                :name: scheme-color-conf-example
 
-               .. code-block:: python
+                html_theme_options = {
+                    "palette": { "scheme": "default" }
+                }
 
-                   html_theme_options = {
-                        "palette": { "scheme": "default" }
-                   }
+            - :test-color-scheme:`default` (scheme for light mode)
+            - :test-color-scheme:`slate` (scheme for dark mode)
 
-            2. The ``slate`` scheme for dark mode.
+            .. only:: html
 
-               .. code-block:: python
+                Click one of the listed schemes to see how it looks.
 
-                   html_theme_options = {
-                        "palette": { "scheme": "slate" }
-                   }
+        .. themeconf:: primary
 
+            The primary color is used for the header, the sidebar, text links and several other
+            components.
 
-        .. themeconf:: primary
+            .. code-block:: python
+                :name: primary-color-conf-example
+
+                html_theme_options = {
+                    "palette": { "primary": "green" }
+                }
 
             Primary color options are
 
-            :red:`red`, :pink:`pink`, :purple:`purple`, :deep-purple:`deep-purple`, :indigo:`indigo`, :blue:`blue`,
-            :light-blue:`light-blue`, :cyan:`cyan`, :teal:`teal`, :green:`green`, :light-green:`light-green`,
-            :lime:`lime`, :yellow:`yellow`, :amber:`amber`, :orange:`orange`, :deep-orange:`deep-orange`,
-            :brown:`brown`, :grey:`grey`, :blue-grey:`blue-grey`, :black:`black`, and :white:`white`.
+            .. jinja:: colors
+
+                {% for color in supported_primary -%}
```

Comment:
How about just using the `raw` directive to directly emit HTML with inline `style` and `data-md-color-primary=xxxx` attributes here?  Then there would be no need to add any CSS styles, and instead you could access the color via `var(--md-primary-fg-color)`.  To support PDF output there could be a separate `.. only:: not html` section emitted as well.  (I think `not html` is allowed in only directives.)

Reply:
You could also emit the inline javascript event handlers, which may be simpler than the separate script you have below.

Reply:
> How about just using the raw directive 

This is originally how I started, but I wanted something that could easily translate (via sphinx native behavior) to other writers (especially non-HTML). I can't simply `app.add_node(name, cls, <visitors>)` in conf.py because the added node's class isn't picklable (due to the scope of the namespace for conf.py).

Reply:
> > How about just using the raw directive
> 
> This is originally how I started, but I wanted something that could easily translate (via sphinx native behavior) to other writers (especially non-HTML). I can't simply `app.add_node(name, cls, <visitors>)` in conf.py because the added node's class isn't picklable (due to the scope of the namespace for conf.py).

For the other builders, though, it will just be a plain list of color names, with no actual colors applied (and obviously no dynamic behavior), so you could just emit plain literals.

I guess to bypass the pickle issue the node could be defined in a separate sphinx extension, e.g. `docs/sphinx_immaterial_docs_ext.py`.

Reply:
> For the other builders, though, it will just be a plain list of color names, with no actual colors applied (and obviously no dynamic behavior), so you could just emit plain literals.

That's why I'm currently using the `docutils.nodes.literal`, just with no syntax highlighting and extra CSS classes. The CSS classes are still used to avoid having to compensate (or compromise) other writers, but maybe it wouldn't matter if the style was inlined.

> I guess to bypass the pickle issue the node could be defined in a separate sphinx extension, e.g. docs/sphinx_immaterial_docs_ext.py.

I was thinking this could be migrated to a badge extension (similar to sphinx-design's badges) in the theme later on. Upstream uses
```html
<button attr=color-value>color value</button>
```
But trying to mimic this from conf.py is how I hit the pickle problem. I like how the button tag automatically changes the mouse cursor, but that seemed like a personal preference.

Reply:
Wouldn't using `nodes.raw` prevent the text from showing up entirely for writers that don't comply with the raw node's given `format` attr?

Reply:
Yes, but then you can provide an alternative version using `only`.

Reply:
Inlining the JS is proving painful because of the curly brackets.

Reply:
The `only` directive can't be nested within a `jinja` directive, so there's still some duplicate code.